src/main/java/edu/harvard/iq/dataverse/util/json/JsonPrinter.java

```diff
@@ -640,18 +640,26 @@ public static JsonObjectBuilder json(MetadataBlock metadataBlock, boolean printO
 
         JsonObjectBuilder fieldsBuilder = Json.createObjectBuilder();
         Set<DatasetFieldType> datasetFieldTypes = new TreeSet<>(metadataBlock.getDatasetFieldTypes());
+
         for (DatasetFieldType datasetFieldType : datasetFieldTypes) {
-            boolean requiredInOwnerDataverse = ownerDataverse != null && ownerDataverse.isDatasetFieldTypeRequiredAsInputLevel(datasetFieldType.getId());
-            boolean displayCondition = !printOnlyDisplayedOnCreateDatasetFieldTypes ||
-                    datasetFieldType.isDisplayOnCreate() ||
-                    requiredInOwnerDataverse;
+            Long datasetFieldTypeId = datasetFieldType.getId();
+            boolean isInputLevel = ownerDataverse != null && ownerDataverse.isDatasetFieldTypeInInputLevels(datasetFieldTypeId);
+            boolean requiredAsInputLevel = isInputLevel && ownerDataverse.isDatasetFieldTypeRequiredAsInputLevel(datasetFieldTypeId);
+            boolean includedAsInputLevel = isInputLevel && ownerDataverse.isDatasetFieldTypeIncludedAsInputLevel(datasetFieldTypeId);
+
+            DatasetFieldType parentDatasetFieldType = datasetFieldType.getParentDatasetFieldType();
+            boolean isRequired = parentDatasetFieldType == null ? datasetFieldType.isRequired() : parentDatasetFieldType.isRequired();
+
+            boolean displayCondition = printOnlyDisplayedOnCreateDatasetFieldTypes
+                    ? (datasetFieldType.isDisplayOnCreate() || isRequired || requiredAsInputLevel)
+                    : !isInputLevel || includedAsInputLevel;
```

Comment:
This is the trickiest one - this is not quite the same as` ownerDataverse==null || includedAsInputLevel`?  - isInputLevel could be false if the ownerDataverse is null or if owner is non-null and the field is not in the inputLevels - do you want to include a field with an owner that is not in the inputLevels (along with ones that are in InputLevels and inputLevel.isInclude() from the second clause)? If the logic is correct as is, I guess the isDatasetFieldTypeInInputLevels is needed, but maybe only has to be called here?

Reply:
The case where ownerDataverse is null is controlled and only occurs when the method is invoked from api/metadatablocks API, not from api/dataverses API. There will be no case where fields associated with a Dataverse and fields not associated with a Dataverse are mixed. It's either one or the other, but not both.

Reply:
Not sure I understand. Line 655 says the (owner is null or the owner exists and there is no inputlevel for this field) or (the owner exists, there is an input level for this field, and the includeLevel  for that entry is true).  Should the (owner exists and there is no inputlevel for this field) condition result in this being true?

Reply:
You are right, sorry. I was not correctly applying the Morgan's laws. Indeed, the logic which I think should be the correct one and I just committed, is: `(ownerDataverse == null || includedAsInputLevelInOwnerDataverse || isNotInputLevelInOwnerDataverse)`. I added the isNotInputLevelInOwnerDataverse condition in the OR statement, which checks that when ownerDataverse exists, there might not be an input level for that field, and therefore it should be shown by default, as it exists in the metadata block. I also renamed the variables for better understanding. Please let me know if you find it appropriate.

Reply:
I think what you're saying means your original code was correct -(owner exists and there is no inputlevel for this field) should result in true (metadata block case), and , unless I'm missing something, the new code has the same logic. I agree the new code is clearer, so I'll go ahead and approve the PR.